man/tcprules.1
.TH tcprules 1
.SH NAME
tcprules \- compiles rules for
.B tcpserver 
or
.BR sslserver .
.SH SYNOPSIS
.B tcprules
.I cdb
.I tmp
.SH DESCRIPTION
.B tcpserver 
and
.B sslserver
optionally follow rules to decide whether a TCP connection is acceptable. 
A 'rule' consists of the tokens 'address', 'instruction', and 'expression' 
telling for which 'address' a 'instruction' is provided following a set of 
optional expressions, providing some environment variables. The 'address' token 
can be given as IPv4/IPv6 address in CIDR format, or 
- if prepended with a equal sign (=) - a hostname/FQDN, to be subject of 
DNS qualification.
Additionally, environment variables can be provided and are promoted for 
a particular connection.  Environment variables may include ':' (colons).

Intructions are
.I allow
and 
.IR deny.
.SH "IPv4 ADDRESSES BASED RULES"
For example, the rule 
.IP
18.23.0.32:deny
.P
prohibits connections from IP address 18.23.0.32.
Ranges of IPv4 addresses can be defined in a \fIclass-dependend\fR manner
.IP
18.:deny
.P 
or by means of a range of contiguous addresses
.IP
18.23.0.1-22:ins
.P
or by using a CIDR notation. The rule
.IP
127./8:allow
.P
accepts any connections from the loopback net. 

\fINote\fR: Always IP addresses with the longest matching prefix
are considered.
.SH "IPv6 ADDRESSES BASED RULES"
.B tcprules
understands \fIcompactified\fR IPv6 addresses in standard
CIDR notation.
The rule
.IP
2001:de01:2:3:4:a:b:c:deny 
.P
rejects any IPv6 packet from a single host while 
.IP
2002::/48:deny
.P
can be used to block an entire IPv6 (sub-)net
in case the net-prefix (here: \fI/48\fR) is provided.

\fINote\fR: Since the IPv6 address on input is evaluated
in its compactified format, simply include the final '::'
for convenience. The resulting address is truncated 
to the specified number of prefix bits. 
.B tcprules
currently does not understand IPv6 mapped IPv4 addresses
.IP
::ffff:a.b.c.d
.P
where 
.IP
a.b.c.d
.P 
is the IPv4 addresses.
In consequence, an IPv4 address will not match a rule for the
corresponding IPv6-mapped address, and vice versa.
.SH USAGE
.B tcprules
reads rules from it's standard input and writes them into
.I cdb
in a binary format suited for quick access by
.BR tcpserver .
Typically
.IP
\fBtcprules\fR rules.cdb rules.tmp < rules.txt
.P

.B tcprules
can be used while
.B tcpserver 
or
.B sslserver
is running. It ensures that
.I cdb
is updated atomically. It does this by first writing the rules to
.I tmp
and then moving
.I tmp
on top of
.IR cdb .
If
.I tmp
already exists, it is destroyed. The directories containing
.I cdb
and
.I tmp
must be writable to
.BR tcprules ;
they must also be on the same filesystem. 

If there is a problem with the input or with
.IR tmp ,
.B tcprules
complains and leaves
.I cdb
alone. 

The binary
.I cdb
format is portable across machines.
.SH RULE FORMAT
A rule is one line. A file containing rules may also contain comments: lines
beginning with # are ignored. 

Each rule contains an address, a colon, an instruction, and optionally 
a comma-separated list of expressions, with no extra spaces. When
.BR tcpserver 
receives a connection from that address, it follows the instruction and evaluates the expressions.
.SH "ADDRESS TOKENS"
.BR tcpserver
looks for rules with various addresses: 
.IP 1.
$TCPREMOTEINFO@$TCPREMOTEIP, if $TCPREMOTEINFO is set; 
.IP 2.
$TCPREMOTEINFO@=$TCPREMOTEHOST, if $TCPREMOTEINFO is set and $TCPREMOTEHOST is
set;
.IP 3.
$TCPREMOTEIP; 
.IP 4.
=$TCPREMOTEHOST, if $TCPREMOTEHOST is set; 
.IP 5.
shorter and shorter prefixes of $TCPREMOTEIP ending with a dot; 
.IP 6.
$TCPREMOTEIP/PREFIX
considering in order the longest matching provided PREFIX;
.IP 7.
shorter and shorter suffixes of $TCPREMOTEHOST starting with a dot, preceded
by =, if $TCPREMOTEHOST is set; 
.IP 8.
=, if $TCPREMOTEHOST is set; and finally 
.IP 9.
the empty string. 
.P
.B tcpserver 
uses the first rule it finds. You should use the
.B -p
option to
.B tcpserver
or
.B sslserver
if you rely on $TCPREMOTEHOST here. 

For example, here are some rules: 

     joe@127.0.0.1:first
     18.23.0.32:second
     :third
     127.:fourth

If $TCPREMOTEIP is 10.119.75.38,
.B tcpserver 
will follow the third instruction. 

If $TCPREMOTEIP is 18.23.0.32,
.B tcpserver
will follow the second instruction. 

If $TCPREMOTEIP is 127.0.0.1 and $TCPREMOTEINFO is bill,
.B tcpserver 
will follow the fourth instruction. 

If $TCPREMOTEIP is 127.0.0.1 and $TCPREMOTEINFO is joe,
.B tcpserver 
will follow the first instruction. 

You can use
.B tcprulescheck
to see how 
.B tcpserver 
will interpret rules in
.IR cdb .
.SH INSTRUCTIONS
The instructions in a rule must be either 
.I allow 
or
.IR deny . 
.I deny 
tells
.B tcpserver 
to drop the connection without running anything. For example, the rule
.IP
:deny
.P
tells
.B tcpserver 
to drop all connections that aren't handled by more specific rules. 

The rule may continue with further expressions used as environment variables, in the form
.IR var="x" .
.B tcpserver
and
.B sslserver
adds an environment variable 
.I $var 
with value
.IR x . 
For example,
.IP
10.0.:allow,RELAYCLIENT="@fix.me"
.P
adds an environment variable $RELAYCLIENT with value @fix.me. The quotes may
be replaced by any repeated character:
.IP
10.0.:allow,RELAYCLIENT=/@fix.me/
.P
Any number of variables may be listed:
.IP
127.0.0.1:allow,RELAYCLIENT="",TCPLOCALHOST="movie.edu"
.IP
.SH "SEE ALSO"
tcpserver(1),
tcprulescheck(1),
argv0(1),
fixcrio(1),
recordio(1),
rblsmtpd(1),
tcpclient(1),
who@(1),
date@(1),
finger@(1),
http@(1),
tcpcat(1),
mconnect(1),
tcp-environ(5)


doc/CHANGES
19960311
	version: ucspi-tcp 0.50, alpha.
19960803
	version: ucspi-tcp 0.60, alpha.
19960903
	version: ucspi-tcp 0.70, beta.
19960914
	version: ucspi-tcp 0.71, beta.
19960922
	version: ucspi-tcp 0.72, beta.
19970410
	version: ucspi-tcp 0.73, beta.
19980118
	version: ucspi-tcp 0.80, beta.
19981107
	code: switched to new install system.
	code: eliminated tcpclient -U; moved usage() to default.
	code: switched to strerr in tcpclient.
	code: switched to str_equal in tcpclient.
	portability problem: OpenBSD connect() prohibits 0.0.0.0.
		impact: tcpclient 0 fails. fix: interpret 0 as
		127.0.0.1. hopefully nobody needs 0 to mean primary
		interface.
	code: tcpclient now interprets empty host name as 0.
	doc: added addcr.0, delcr.0.
	doc: added who@.0, date@.0, finger@.0.
	doc: added tcpcat.0, mconnect.0.
	code: added printstatus() in tcpserver.
	code: eliminated numchildren in printpid() in tcpserver.
	code: unblocked SIGCHLD after setting SIGCHLD to SIG_DFL in
		tcpserver, instead of before, to avoid confusion if
		someone sends SIGCHLD manually.
	code: moved printstatus() before fork in tcpserver.
	code: added mconnect-io.
	code: added argv0.
	code: added fixcr.
	doc: revamped BLURB.
	code: added recordio.
	doc: cleaned up tcpclient.0.
	code: added http@.
	code: stopped after 100 bytes in safeappend() in tcpserver.
19981108
	code: added tcpserver -B.
	code: added a few regression tests.
	version: ucspi-tcp 0.83, beta.
19981109
	doc: added http@ to BLURB.
19981110
	code: added exec to mconnect.
	code: tcpclient now ignores port results from getpeername() in
		favor of the port it tried connecting to.
	code: added strport to tcpclient.
	doc: reorganized tcpclient.0.
	doc: reorganized tcpserver.0.
	code: revamped tcpserver messages.
	code: revamped tcpclient messages.
	code: added a few more regression tests.
	code: rewrote recordio for instant output and other features.
	doc: added references in tcpserver.0 and tcpclient.0.
	code: added tcpclient -i option.
	code: added tcprulescheck.
	code: added some tcprules regression tests.
19981111
	code: avoided ndelay in recordio.
	version: ucspi-tcp 0.84, beta.
19981212
	doc: improved description of host 0 in tcpserver.0. tnx to
		several people.
19981218
	doc: reordered sample rules in tcprules.0.
		tnx Harald Hanche-Olsen.
20000120
	internal: revamped packaging.
	internal: massive rewrite. tnx to everyone for bug reports and
		other comments on various versions of tcpclient,
		tcpserver, tcpcontrol, and rblsmtpd: Akihiro Iijima,
		Akihiro Terasaki, Albert J. deVera, Alex Vostrikov,
		Amos Shapira, Anand R. Buddhdev, Andrea Paolini,
		Andrew Pam, Araki Yasuhiro, Arne Wichmann,
		Ayamura Kikuchi, Bart B. B. Hanssens, Bert Gijsbers,
		Bradford M. Shelton, Brendan Kehoe, Brian J. Reichert,
		Brian M. Fisk, Brian O'Reilly, Bruno Wolff,
		Chris Garrigues, Chris Johnson, Christian Wettergren,
		Dale N. Woolridge, Dan M. Vogel, Daniel C. Mahoney,
		Darren Hall, Darren W. Rees, Dave Sill, David J. Walton,
		David P. Smith, David Pool, Dax Kelson, Dirk Jaeckel,
		Dirk Vluegels, Donald E. Blais, Ed Weinberg,
		Edward S. Marshall, Eric A. Perlman, Erik Wallin,
		Faried Nawaz, Fred B. Ringel, Frederik P. Lindberg,
		Gerry Boudreaux, Giles Lean, Grant Holliday,
		Greg D. Patterson, Harald Barth, Harald Fritzsche,
		Harald Hanche-Olsen, Hirofumi Ukawa, Hiroshi Yamashita,
		Hitesh Patel, Ingmar Hupp, J. B. Keith Humphreys,
		Jakub K. Boguslaw, Janos Farkas, Jason R. Mastaler,
		Jeff Hayward, Jeremy Wohl, Jim Littlefield,
		Johan Holmberg, John Bolhuis, John D. Mitchell,
		Jos Backus, Jose Monteiro, Joshua J. Ellis,
		Julie L. Baumler, Karl Lehenbauer, Karsten Thygesen,
		Keith Burdis, Kenny Elliott, Kikuchi Kousuke,
		Kris Kennaway, Lars Balker Rasmussen, Louis S. Theran,
		Lyndon F. Bartels, Mads E. Eilertsen, Mark Delany,
		Martin Mersberger, Matt P. Simerson, Matthew A. Zahorik,
		Michael B. Scher, Michael Handler, Michael Hirohama,
		Michael R. Gile, Michael Salmon, Mikael Suokas,
		Nobuhiro Murata, Patrick M. Kane, Paul R. Rotering,
		Peter Rye, Peter Wilkinson, Petr Novotny,
		Petri Kaukasoina, Raul D. Miller, Richard A. Soderberg,
		Robert W. Luce, Roberto A. Lumbreras, Russ Allbery,
		Russell Nelson, Sean Reifschneider, Shawn A. Clifford,
		Shin Ohira, Shinya O'Hira, Stan Norton,
		Stathy G. Touloumis, Stefan M. Linnemann, Sudish Joseph,
		Thomas E. Erskine, Thomas Kuerten, Tim Goodwin,
		Timothy L. Mayo, Todd Underwood, Tomoaki Terazawa,
		Tomoki Yoshioka, Toshinori Maeno, Uwe Ohse, Vern Hart,
		Vince Vielhaber, Waskita Adijarto, William E. Baxter,
		Wolfgang Rufeger, Wu Ching-hong, Yoshitatsu Takeshita.
20000307
	internal: switched to various new libraries.
	ui: tcpserver prints status on startup, before first connection.
20000309
	ui: incorporated rblsmtpd.
	ui: tcpserver -X allows nonexistent rules file.
	ui: http@ uses HTTP/1.0, sends Host, removes header.
	ui: added fixcrio.
	ui: delcr leaves CR alone at the end of a partial final line.
20000311
	ui: switched to prot; so setgid() is preceded by setgroups().
	ui: tcpserver supports -U.
	ui: tcpserver supports hostname rules.
	ui: tcprulescheck now uses environment variables.
20000312
	version: ucspi-tcp 0.86, beta.
20000314
	portability problem: the poll() emulation in RedHat 5.1 doesn't
		clear revents when select() returns 0. tnx Petr Novotny.
		impact: dns lookups busy-loop when they should time out.
		fix: new iopause from DNScache.
20000315
	version: ucspi-tcp 0.87, beta.
20000318
	internal: split rules() out of tcpserver and tcprulescheck.
	bug: didn't always clear rule length when $TCPREMOTEHOST was
		set. impact: empty rule could be ignored. fix: obvious.
		tnx Toshinori Maeno, Takashi Takizawa, Yuichi Katoh.
	ui: check for lone = rule if $TCPREMOTEHOST is set.
	ui: tcpclient supports fast+slow timeouts, 2+58 by default.
	ui: tcpclient randomizes order of addresses.
	version: ucspi-tcp 0.88, beta.
20080227 (feh)
	added Greetdelay option into rblsmtpd.

src/tcprules.c
#include <unistd.h>
#include "logmsg.h"
#include "stralloc.h"
#include "getln.h"
#include "buffer.h"
#include "exit.h"
#include "fmt.h"
#include "byte.h"
#include "cdbmake.h"
#include "ip.h"
#include "scan.h"
#include "open.h"
#include "ip_bit.h"

#define WHO "tcprules"

extern int rename(const char *,const char *); /* declared in stdio, ugh */

unsigned long linenum = 0;
char *fntemp;
char *fn;

stralloc line = {0};
int match = 1;

stralloc address = {0};
stralloc ipstring = {0};
stralloc data = {0};
stralloc key = {0};

struct cdb_make c;

void nomem(void) {
  logmsg(WHO,111,FATAL,"out of memory");
}

void usage(void) {
  logmsg(WHO,100,USAGE,"tcprules rules.cdb rules.tmp");
}

void die_bad(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("unable to parse this line: ",line.s));
}

void die_write(void) {
  logmsg(WHO,111,FATAL,B("unable to write to: ",fntemp));
}

void die_length(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("invalid prefix length on line: ",line.s));
}

void die_ip(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("invalid address on line: ",line.s));
}
void die_ip4c(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("invalid IPv4 CIDR address on line: ",line.s));
}
void die_ip6(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("invalid IPv6 address: on line: ",line.s));
}
void die_ip6c(void) {
  if (!stralloc_0(&line)) nomem();
  logmsg(WHO,101,SYNTAX,B("invalid IPv6 CIDR address on line: ",line.s));
}

char strnum[FMT_ULONG];
stralloc sanum = {0};

void getnum(char *buf, int len, unsigned long *u) {
  if (!stralloc_copyb(&sanum,buf,len)) nomem();
  if (!stralloc_0(&sanum)) nomem();
  if (sanum.s[scan_ulong(sanum.s,u)]) die_bad();
}

void doaddressdata(void) {
  int i; 
  int ipv6 = 0;
  int cidr = 0;
  int info = 0;
  int left;
  int right;
  unsigned long bot;
  unsigned long top;
  unsigned long prefix;
  stralloc ip6address = {0};

  if (byte_chr(address.s,address.len,'=') < address.len) goto CDB;
  if (byte_chr(address.s,address.len,':') < address.len) ipv6 = 1;
  if (byte_chr(address.s,address.len,'/') < address.len) cidr = 1;
  if (byte_chr(address.s,address.len,'@') < address.len) info = 1;

  if (info) {
    if (ipv6 && !cidr) {
      i = byte_chr(address.s,address.len,'@');
      if (!stralloc_copyb(&key,address.s,i)) nomem();
      if (!stralloc_cats(&key,"@")) nomem();
      if (!stralloc_copyb(&ipstring,address.s + i + 1,address.len - i - 1)) nomem();
      if (!stralloc_0(&ipstring)) nomem();
      if (!ip6_fmt_str(&ip6address,ipstring.s)) {
        if (!stralloc_catb(&key,ip6address.s,ip6address.len)) nomem();
        if (cdb_make_add(&c,key.s,key.len,data.s,data.len) == -1) die_write();
        return;
      }
      else 
        die_ip6();
    }
    goto CDB;
  }
  
  if (!ipv6) {
    if (cidr) {
      i = byte_chr(address.s,address.len,'/');
      getnum(address.s + i + 1,address.len - i - 1,&prefix);
      if (prefix > 32 || prefix <= 0) die_length();
      address.s[i] = '\0';

      switch (ip4_bitstring(&ipstring,address.s,prefix)) {
        case -1: nomem();
        case  0: if (!stralloc_copys(&key,"_")) nomem();
                 if (!stralloc_catb(&key,ipstring.s,ipstring.len)) nomem();
                 if (cdb_make_add(&c,key.s,key.len,data.s,data.len) == -1) die_write();
                 break;
        case  1: die_ip4c();
      }
      return;
    }
    else if ((i = byte_chr(address.s,address.len,'-')) < address.len) {	
      left = byte_rchr(address.s,i,'.');
      if (left == i) left = 0; else ++left; 
            
      ++i;
      right = i + byte_chr(address.s + i,address.len - i,'.');
            
      getnum(address.s + left,i - 1 - left,&bot);
      getnum(address.s + i,right - i,&top);
      if (top > 255) top = 255;
      while (bot <= top) {
        if (!stralloc_copyb(&key,address.s,left)) nomem();
        if (!stralloc_catb(&key,strnum,fmt_ulong(strnum,bot))) nomem();
        if (!stralloc_catb(&key,address.s + right,address.len - right)) nomem();
        if (cdb_make_add(&c,key.s,key.len,data.s,data.len) == -1) die_write();
        ++bot;
      }
      return;
    }
  }
 
  if (ipv6) {
    if (cidr) {
      i = byte_chr(address.s,address.len,'/');
      getnum(address.s + i + 1,address.len - i - 1,&prefix);
      if (prefix > 128 ) die_length();
      switch (ip6_bitstring(&ipstring,address.s,prefix)) {
        case -1: nomem();
        case  0: if (!stralloc_copys(&key,"^")) nomem();
                 if (!stralloc_catb(&key,ipstring.s,ipstring.len)) nomem();
                 if (cdb_make_add(&c,key.s,key.len,data.s,data.len) == -1) die_write();
                 break;
        case  1: die_ip6c();
      }
      return;
    }
    else {
      if (!stralloc_copyb(&ipstring,address.s,address.len)) nomem();
      if (!stralloc_0(&ipstring)) nomem();
      switch (ip6_fmt_str(&ip6address,ipstring.s)) {
        case -1: nomem();
        case  0: if (cdb_make_add(&c,ip6address.s,ip6address.len,data.s,data.len) == -1)
                   die_write();
                 break;
        case  1: die_ip6();
      }
      return;
    }
  }

  CDB:
  if (cdb_make_add(&c,address.s,address.len,data.s,data.len) == -1) die_write();
  return;
}

int main(int argc,char **argv) {
  int colon;
  char *x;
  int len;
  int fd;
  int i;
  char ch;

  fn = argv[1];
  if (!fn) usage();
  fntemp = argv[2];
  if (!fntemp) usage();

  fd = open_trunc(fntemp);
  if (fd == -1)
    logmsg(WHO,111,ERROR,B("unable to create: ",fntemp));
  if (cdb_make_start(&c,fd) == -1) die_write();

  while (match) {
    if (getln(buffer_0,&line,&match,'\n') == -1)
      logmsg(WHO,111,FATAL,"unable to read input");

     x = line.s;
     len = line.len;

     if (!len) break;
     if (x[0] == '#') continue; 	/* ignore comments */
     if (x[0] == '\n') continue;	/* ignore empty lines */

     while (len) {
       ch = x[len - 1];
       if (ch != '\n') if (ch != ' ') if (ch != '\t') break;
       --len;
     }
     line.len = len; 			/* for die_bad() */

/* Need to distinguish between IPv(4|6) address and FQDN ... */
  
    i = byte_chr(x,len,',');		/* environment variables start */
    colon = byte_rchr(x,i,':');

    if (colon == len) continue;

    if (!stralloc_copyb(&address,x,colon)) nomem();
    if (!stralloc_copys(&data,"")) nomem();

    x += colon + 1; len -= colon + 1;

    if ((len >= 4) && byte_equal(x,4,"deny")) {
      if (!stralloc_catb(&data,"D",2)) nomem();
      x += 4; len -= 4;
    }
    else if ((len >= 5) && byte_equal(x,5,"allow")) {
      x += 5; len -= 5;
    }
    else 
      die_bad();

    while (len)
      switch (*x) {
        case ',': i = byte_chr(x,len,'=');
                  if (i == len) die_bad();
                  if (!stralloc_catb(&data,"+",1)) nomem();
                  if (!stralloc_catb(&data,x + 1,i)) nomem();
                  x += i + 1; len -= i + 1;
                  if (!len) die_bad();
                  ch = *x;
                  x += 1; len -= 1;
                  i = byte_chr(x,len,ch);
                  if (i == len) die_bad();
                  if (!stralloc_catb(&data,x,i)) nomem();
                  if (!stralloc_0(&data)) nomem();
                  x += i + 1; len -= i + 1;
                  break;
         default: die_bad();
       }

    doaddressdata();
  }

  if (cdb_make_finish(&c) == -1) die_write();
  if (fsync(fd) == -1) die_write();
  if (close(fd) == -1) die_write(); /* NFS stupidity */
  if (rename(fntemp,fn))
    logmsg(WHO,111,ERROR,B("unable to move ",fntemp," to: ",fn));

  _exit(0);
}
